backend/pop3/src/main/java/com/fsck/k9/backend/pop3/Pop3Sync.kt

```diff
@@ -0,0 +1,660 @@
+
+            Timber.v("SYNC: About to get local folder %s", folder)
+            backendFolder = backendStorage.getFolder(folder)
+
+            listener.syncStarted(folder)
+
+            /*
+             * Get the message list from the local store and create an index of
+             * the uids within the list.
+             */
+            var localUidMap: Map<String, Long?> = backendFolder.getAllMessagesAndEffectiveDates()
+
+            Timber.v("SYNC: About to get remote folder %s", folder)
+            remoteFolder = remoteStore.getFolder(folder)
+
+            /*
+             * Synchronization process:
+             *
+            Open the folder
+            Upload any local messages that are marked as PENDING_UPLOAD (Drafts, Sent, Trash)
+            Get the message count
+            Get the list of the newest K9.DEFAULT_VISIBLE_LIMIT messages
+            getMessages(messageCount - K9.DEFAULT_VISIBLE_LIMIT, messageCount)
+            See if we have each message locally, if not fetch it's flags and envelope
+            Get and update the unread count for the folder
+            Update the remote flags of any messages we have locally with an internal date newer than the remote message.
+            Get the current flags for any messages we have locally but did not just download
+            Update local flags
+            For any message we have locally but not remotely, delete the local message to keep cache clean.
+            Download larger parts of any new messages.
+            (Optional) Download small attachments in the background.
+             */
+
+            /*
+             * Open the remote folder. This pre-loads certain metadata like message count.
+             */
+            Timber.v("SYNC: About to open remote folder %s", folder)
+
+            remoteFolder.open()
+
+            listener.syncAuthenticationSuccess()
+
+            /*
+             * Get the remote message count.
+             */
+            val remoteMessageCount = remoteFolder.messageCount
+
+            var visibleLimit = backendFolder.visibleLimit
+
+            if (visibleLimit < 0) {
+                visibleLimit = syncConfig.defaultVisibleLimit
+            }
+
+            val remoteMessages: MutableList<Pop3Message?> = ArrayList<Pop3Message?>()
+            val remoteUidMap: MutableMap<String?, Pop3Message?> = HashMap<String?, Pop3Message?>()
+
+            Timber.v("SYNC: Remote message count for folder %s is %d", folder, remoteMessageCount)
+
+            val earliestDate = syncConfig.earliestPollDate
+            val earliestTimestamp = if (earliestDate != null) earliestDate.time else 0L
+
+            /* Message numbers start at 1. */
+            var remoteStart = 1
+            if (remoteMessageCount > 0) {
+                // Adjust the starting message number based on the visible limit
+                if (visibleLimit > 0) {
+                    remoteStart += max(0, remoteMessageCount - visibleLimit)
```

Comment:
Suggestion: I find Kotlin's `coerceAtLeast()` makes code like this easier to read.

```suggestion
                    remoteStart += (remoteMessageCount - visibleLimit).coerceAtLeast(0)
```